### PR TITLE
fix(web): Add BkSp1 special characer

### PR DESCRIPTION
**Pull request by MayuraVerma**
added 
'*BkSp1*': 0x21
to resolve https://github.com/keymanapp/keyman/issues/10220

### CODE REVIEW

**Comment by keyman-server** [comment]
This pull request is from an external repo and will not automatically be built. The build must still be passed before it can be merged. Ask one of the team members to make a manual build of this PR.

File: web/src/engine/osk/src/specialCharacters.ts
Comment (darcywong00) [inline web/src/engine/osk/src/specialCharacters.ts:21]: Thanks for your submission! Could you move this block between lines 48-49?
That's the section of new characters added in Keyman 17.0.

At some point, we'd also want to update Developer to add this character. (Similar to #9846)

Comment (darcywong00) [reply web/src/engine/osk/src/specialCharacters.ts:21]: It would be also nice to update the test keyboard
web/src/test/manual/web/issue9469/test9469/source/test9469.keyman-touch-layout

and add a key with text `*BkSp1*` to test the new backspace keycap.

**Review by jahorton — CHANGES_REQUESTED** [review]
Thanks for submitting a pull request!

Unfortunately, these changes only cover a _part_ of the problem you're seeking to address, rather than all of it.  Fortunately, the hardest part appears to already be in place - I can confirm that the symbol you're seeking to use does, in fact, exist already within the relevant font.

(For other reviewers: refer to https://github.com/silnrsi/font-symchar/blob/main/documentation/encoding.md, CTRL-F `U+E021`.  I believe it's a private repository, so apologies to external contributors.)

That said, note that this change does not selectively choose between `'*BkSp*'` and `'*BkSp1*'` when relevant - we'd want to pick the 'right' BkSp for the device, rather than make things mismatch worse on some.  That said, it does appear that modern iOS and Android devices (I quick-tested using an API 32 emulated device) both use something notably closer to your `'*BkSp1*'` than the existing `'*BkSp*'`.

Checking against Android API 21, that Android version's Gboard _does_ use something closer to `'*BkSp*'`, though.  I'm not sure where the transition point for this is... and/or how high priority we consider keeping the old style in place for older devices such as this one.

As this submission stands, current keyboards would see no change.  They'd have to explicitly specify use of the alternate backspace style within their touch-layout to see any difference.  We'd need to replace `'*BkSp*'`'s value directly to see the difference.  I'm not the final judge to convince on this, but I think it'd actually be reasonable to make `0x21` the default BkSp glyph and `0x04` a 'legacy' version.

From that, we'd _ideally_ have a way to condition on the Android version and auto-select the best-matching version on older Android devices... something that would require extra code.  Not sure if it's something we'd highly-prioritize, though.

@mcdurdin likes to pitch in on design / styling decisions like this, so I'm pinging him on this in hopes of getting a more definitive answer and/or direction to take here.  There's a chance that his decision might depend upon when the Gboard transition-point is - on how long ago Android's change to the BKSP glyph occurred, and that may take a bit of extra research to determine.

**Comment by mcdurdin** [comment]
Thank you for your contribution.

I think this is not something we want to address in this way. It needs to be addressed with user-side theming support:

* On Android, the 'stock' keyboard is dependent on the vendor, and there is a wide range of different styles. On iOS, of course the 'stock' keyboard is always Apple. However, as I noted in #10226, we don't actually have a specific goal to precisely mimic keyboard styling from the stock platform.

* This should also not be a keyboard designer's styling decision, but rather something that is styling specific to the platform. So adding `*BkSp1*` is actually heading in the wrong direction.

* Furthermore, we don't currently have a glyph for the RTL backspace, outlined. So we'd need to add that before we could merge this in any case.

So for those reasons, I don't think we can move ahead with this, sorry.

**Comment by MayuraVerma** [comment]
@mcdurdin Please let me know if I could work on any specific simple tasks. I can contribute to this project in little ways.

**Comment by mcdurdin** [comment]
Hi @MayuraVerma, thank you for your contributions. They are always appreciated. One big area where we could really use help is in the documentation. Even looking at individual topics in the Developer documentation and improving the wording is helpful.

We have a few issues listed in this repo as ['help wanted'](https://github.com/keymanapp/keyman/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22) and I know that you looked at for example #2655 relating to this specific PR, and we said we welcome pull requests in that issue, so this I appreciate may be a little frustrating.

We're currently holding off as far as possible on minor tweaks and improvements to the on screen keyboard, because each tweak represents another thing we have to factor into the upcoming theming rewrite. It's already tricky because a number of keyboards have their own styling and I don't know how we are going to integrate that with a more general theming system!

We'll keep brainstorming here for other "low hanging fruit" issues that you could take on.

**Comment by MayuraVerma** [comment]
> Thank you for your contribution.
> 
> I think this is not something we want to address in this way. It needs to be addressed with user-side theming support:
> 
> * On Android, the 'stock' keyboard is dependent on the vendor, and there is a wide range of different styles. On iOS, of course the 'stock' keyboard is always Apple. However, as I noted in [feat(ios): How to achieve iOS stock keyboard backspace behavior #10226](https://github.com/keymanapp/keyman/issues/10226), we don't actually have a specific goal to precisely mimic keyboard styling from the stock platform.
> * This should also not be a keyboard designer's styling decision, but rather something that is styling specific to the platform. So adding `*BkSp1*` is actually heading in the wrong direction.
> * Furthermore, we don't currently have a glyph for the RTL backspace, outlined. So we'd need to add that before we could merge this in any case.
> 
> So for those reasons, I don't think we can move ahead with this, sorry.

@mcdurdin 
Although the symbol font has glyph for U+E021, keyman is not displaying the text, but a TOFU.

There is no documentation of embedded css/js to make this work.

Please help

Screenshot 2023-12-16 at 4 40 44 PM

image


OR

1. I will update the font with right hand version of glyph. I will close this issue. 
https://github.com/silnrsi/font-symchar/issues/2

2. I will update this documentation
https://help.keyman.com/developer/16.0/guides/develop/creating-a-touch-keyboard-layout-for-amharic-the-nitty-gritty

image

with a row indicating alternate glyph.

If it helps accepting this pull request.

Thank you.

**Comment by mcdurdin** [comment]
> Although the symbol font has glyph for U+E021, keyman is not displaying the text, but a TOFU.
> 
> There is no documentation of embedded css/js to make this work.

The E021 codepoint is internal and subject to change, so keyboards should not make use of it. The codepoints can and may well change between Keyman versions. Thus, this is undocumented.

**Pull request closed without merging**
